### PR TITLE
add test for staking/register handler 2021

### PR DESCRIPTION
**Pull request by lzxm160**
work on #2021

### CODE REVIEW

File: action/protocol/staking/handlers_test.go [hunk 1]
```diff
@@ -166,6 +167,1624 @@ func TestProtocol_HandleCreateStake(t *testing.T) {
...
+				require.Equal(ErrNilParameters, errors.Cause(err))
+				candidate, err = getCandidate(sm, test.caller)
+				require.NoError(err)
+				require.Equal(test.caller.String(), candidate.Owner.String())
+			} else {
+				require.NotNil(candidate)
+				require.NoError(err)
+				require.Equal(test.ownerAddrStr, candidate.Owner.String())
+			}
+			require.Equal("1", candidate.Votes.String())
```
Comment (dustinxie) [inline action/protocol/staking/handlers_test.go]: test.amountStr, instead of "1"

File: action/protocol/staking/handlers_test.go [hunk 2]
```diff
@@ -166,6 +167,1624 @@ func TestProtocol_HandleCreateStake(t *testing.T) {
...
+			candidate = p.inMemCandidates.GetByOwner(candidate.Owner)
+			require.NotNil(candidate)
+			require.Equal("1", candidate.Votes.String())
```
Comment (dustinxie) [inline action/protocol/staking/handlers_test.go]: same as above

File: action/protocol/staking/handlers_test.go [hunk 3]
```diff
@@ -166,6 +167,1624 @@ func TestProtocol_HandleCreateStake(t *testing.T) {
...
+				require.Equal(ErrNilParameters, errors.Cause(err))
+				candidate, err = getCandidate(sm, test.caller)
+				require.NoError(err)
+				require.Equal(test.caller.String(), candidate.Owner.String())
+			} else {
+				require.NotNil(candidate)
+				require.NoError(err)
+				require.Equal(test.ownerAddrStr, candidate.Owner.String())
+			}
+			require.Equal("158", candidate.Votes.String())
```
Comment (dustinxie) [inline action/protocol/staking/handlers_test.go]: same as above

File: action/protocol/staking/handlers_test.go [hunk 4]
```diff
@@ -166,6 +167,1624 @@ func TestProtocol_HandleCreateStake(t *testing.T) {
...
+			if test.updateName != "" {
+				require.Equal(test.updateName, candidate.Name)
+			}
+			if test.updateOperator != "" {
+				require.Equal(test.updateOperator, candidate.Operator.String())
+			}
+			if test.updateOperator != "" {
+				require.Equal(test.updateReward, candidate.Reward.String())
+			}
+			require.Equal("158", candidate.Votes.String())
```
Comment (dustinxie) [inline action/protocol/staking/handlers_test.go]: same as above

File: action/protocol/staking/handlers_test.go [hunk 5]
```diff
@@ -166,6 +167,1624 @@ func TestProtocol_HandleCreateStake(t *testing.T) {
...
+			bucket, err := getBucket(sm, indices[0])
+			require.NoError(err)
+			require.Equal(candidate.Owner.String(), bucket.Candidate.String())
+			require.Equal(test.caller.String(), bucket.Owner.String())
+			require.Equal(test.amount, bucket.StakedAmount.String())
+
+			// test candidate
+			candidate, err = getCandidate(sm, candidate.Owner)
+			require.NoError(err)
+			require.Equal("2", candidate.Votes.String())
```
Comment (dustinxie) [inline action/protocol/staking/handlers_test.go]: same as above

File: action/protocol/staking/handlers_test.go [hunk 6]
```diff
@@ -166,6 +167,1624 @@ func TestProtocol_HandleCreateStake(t *testing.T) {
...
+			require.NoError(err)
+			require.Equal(identityset.Address(1).String(), bucket.Candidate.String())
+			require.Equal(identityset.Address(1).String(), bucket.Owner.String())
+			require.Equal(test.amount, bucket.StakedAmount.String())
+
+			// test candidate
+			candidate, err := getCandidate(sm, candidate.Owner)
+			require.NotNil(candidate)
+			require.NoError(err)
+			require.Equal("20000000000000000002", candidate.Votes.String())
```
Comment (dustinxie) [inline action/protocol/staking/handlers_test.go]: initBalance + xxx?

File: action/protocol/staking/handlers_test.go [hunk 7]
```diff
@@ -166,6 +167,1624 @@ func TestProtocol_HandleCreateStake(t *testing.T) {
...
+			require.Equal(test.candidateName, candidate.Name)
+			require.Equal(candidate.Operator.String(), candidate.Operator.String())
+			require.Equal(candidate.Reward.String(), candidate.Reward.String())
+			require.Equal(candidate.Owner.String(), candidate.Owner.String())
+			require.Equal("1200000000000000000000000", candidate.SelfStake.String())
```
Comment (dustinxie) [inline action/protocol/staking/handlers_test.go]: test.xxx?

File: action/protocol/staking/handlers_test.go [hunk 8]
```diff
@@ -166,6 +167,1624 @@ func TestProtocol_HandleCreateStake(t *testing.T) {
...
+			candidate = p.inMemCandidates.GetByOwner(candidate.Owner)
+			require.NotNil(candidate)
+			require.Equal("20000000000000000002", candidate.Votes.String())
```
Comment (dustinxie) [inline action/protocol/staking/handlers_test.go]: same

File: action/protocol/staking/handlers_test.go [hunk 9]
```diff
@@ -166,6 +167,1624 @@ func TestProtocol_HandleCreateStake(t *testing.T) {
...
+			10000,
+			false,
+			true,
+			nil,
+			iotextypes.ReceiptStatus_Success,
+		},
+	}
+
+	for _, test := range tests {
+		if test.newProtocol {
```
Comment (dustinxie) [inline action/protocol/staking/handlers_test.go]: why have this? to clear the sm?

File: action/protocol/staking/handlers_test.go [hunk 10]
```diff
@@ -166,6 +167,1624 @@ func TestProtocol_HandleCreateStake(t *testing.T) {
...
+			sm, p, candidate, _ = initAll(t, ctrl)
+		} else {
+			candidate = candidate2
+		}
+		ctx, createCost := initCreateStake(t, sm, test.caller, test.initBalance, test.gasPrice, test.gasLimit, test.nonce, test.blkHeight, test.blkTimestamp, test.blkGasLimit, p, candidate, test.amount, false)
+		act, err := action.NewUnstake(test.nonce, test.index,
+			nil, test.gasLimit, test.gasPrice)
+		require.NoError(err)
+		if test.clear {
```
Comment (dustinxie) [inline action/protocol/staking/handlers_test.go]: why delete caller? if test.caller runs an action, it should be kept in inMemCandidates?

File: action/protocol/staking/handlers_test.go (hunk 6 above)
Comment (lzxm160) [reply action/protocol/staking/handlers_test.go]: votes is 2 when call initAll,I will change to 0 in next pr

File: action/protocol/staking/handlers_test.go (hunk 9 above)
Comment (lzxm160) [reply action/protocol/staking/handlers_test.go]: for test inMemCandidates.GetByOwner,this error cannot produce if dont delete candidate in "inMemCandidates" cache

File: action/protocol/staking/handlers_test.go (hunk 10 above)
Comment (lzxm160) [reply action/protocol/staking/handlers_test.go]: for test inMemCandidates.GetByOwner,this error cannot produce if dont delete candidate in "inMemCandidates" cache

File: action/protocol/staking/handlers_test.go (hunk 9 above)
Comment (lzxm160) [reply action/protocol/staking/handlers_test.go]: right,this flag is for "fetchBucket, bucket.Owner is not equal to actionCtx.Caller" error.
That can only produce in the same sm

File: action/protocol/staking/handlers_test.go [hunk 11]
```diff
@@ -751,10 +763,10 @@ func TestProtocol_HandleUnstake(t *testing.T) {
 			// test candidate
 			candidate, err = getCandidate(sm, candidate.Owner)
 			require.NoError(err)
-			require.Equal("2", candidate.Votes.String())
+			require.Equal(test.afterUnstake, candidate.Votes.String())
```
Comment (dustinxie) [inline action/protocol/staking/handlers_test.go]: the value should be 2, why it's not "2" added in above?

**Pull request merged**